### PR TITLE
[SPARK-41366][CONNECT] DF.groupby.agg() should be compatible

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
This patch fixes small inconsistencies between the PySpark API and Spark Connect.

### Why are the changes needed?
Compatibility

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

File: python/pyspark/sql/connect/dataframe.py [hunk 1]
```diff
@@ -55,8 +60,109 @@ def __init__(self, df: "DataFrame", *grouping_cols: Union[Column, str]) -> None:
...
+        |Alice|           2|
+        |  Bob|           5|
+        +-----+------------+
+        """
+        assert exprs, "exprs should not be empty"
+        if len(exprs) == 1 and isinstance(exprs[0], dict):
+            from pyspark.sql.connect.function_builder import functions as FB
+
+            # Convert the dict into key value pairs
+            measures = [Column(ScalarFunctionExpression(exprs[0][k], col(k))) for k in exprs[0]]
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/dataframe.py:160]: I think this PR needs rebasing now.

Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:160]: Do you mean anything in particular? I had just built this on top of recent master.

Comment (zhengruifeng) [reply python/pyspark/sql/connect/dataframe.py:160]: `ScalarFunctionExpression` was just changed in https://github.com/apache/spark/commit/70ab135813fd224fdc80ab30622e57399a3a49f2

Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:160]: ok :( will do

File: python/pyspark/sql/connect/dataframe.py [hunk 2]
```diff
@@ -55,8 +59,109 @@ def __init__(self, df: "DataFrame", *grouping_cols: Union[Column, str]) -> None:
...
+               and certain groups are too large to fit in memory.
+
+           .. seealso:: :func:`pyspark.sql.functions.pandas_udf`
+
+        If ``exprs`` is a single :class:`dict` mapping from string to string, then the key
+        is the column to perform aggregation on, and the value is the aggregate function.
+
+        Alternatively, ``exprs`` can also be a list of aggregate :class:`Column` expressions.
+
+        .. versionadded:: 1.3.0
```
Comment (amaliujia) [inline python/pyspark/sql/connect/dataframe.py:91]:
```suggestion
        .. versionadded:: 3.4.0
```

File: python/pyspark/sql/connect/dataframe.py [hunk 3]
```diff
@@ -55,8 +59,109 @@ def __init__(self, df: "DataFrame", *grouping_cols: Union[Column, str]) -> None:
...
+        +-----+------------+
+        | name|min_udf(age)|
+        +-----+------------+
+        |Alice|           2|
+        |  Bob|           5|
+        +-----+------------+
+        """
+        assert exprs, "exprs should not be empty"
+        if len(exprs) == 1 and isinstance(exprs[0], dict):
+            from pyspark.sql.connect.function_builder import functions as FB
```
Comment (amaliujia) [inline python/pyspark/sql/connect/dataframe.py:156]: This import is not used?

File: python/pyspark/sql/connect/dataframe.py [hunk 4]
```diff
@@ -55,8 +59,109 @@ def __init__(self, df: "DataFrame", *grouping_cols: Union[Column, str]) -> None:
...
+
+            # Convert the dict into key value pairs
+            measures = [scalar_function(exprs[0][k], col(k)) for k in exprs[0]]
+        else:
+            # Columns
+            assert all(isinstance(c, Column) for c in exprs), "all exprs should be Column"
```
Comment (amaliujia) [inline python/pyspark/sql/connect/dataframe.py:162]: Nit:

Do we need to put the full package name into the error message given that we have two Columns: spark.sql.column and spark.sql.connect.column? 

Do you foresee when users use connect client, will they be confused on two columns?

File: python/pyspark/sql/connect/dataframe.py [hunk 5]
```diff
@@ -55,8 +59,109 @@ def __init__(self, df: "DataFrame", *grouping_cols: Union[Column, str]) -> None:
...
+        +-----+--------+
+        | name|min(age)|
+        +-----+--------+
+        |Alice|       2|
+        |  Bob|       5|
+        +-----+--------+
+
+        Same as above but uses pandas UDF.
+
+        >>> @pandas_udf('int', PandasUDFType.GROUPED_AGG)  # doctest: +SKIP
```
Comment (amaliujia) [inline python/pyspark/sql/connect/dataframe.py:142]: Actually maybe should modify this: this is not supported by Connect python client yet....

**Review by zhengruifeng — APPROVED** [review]
LGTM pending the python linter

File: python/pyspark/sql/connect/dataframe.py (hunk 3 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:156]: Done

File: python/pyspark/sql/connect/dataframe.py (hunk 4 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:162]: Right now this is the literal code from PySpark, we probably have to revisit this once Hyukjins patch is further along

File: python/pyspark/sql/connect/dataframe.py (hunk 5 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:142]: removed

File: python/pyspark/sql/connect/dataframe.py (hunk 2 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:91]: Done

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 6]
```diff
@@ -397,7 +397,7 @@ class SparkConnectPlanner(session: SparkSession) {
   }
 
   private def transformUnresolvedExpression(exp: proto.Expression): UnresolvedAttribute = {
-    UnresolvedAttribute(exp.getUnresolvedAttribute.getUnparsedIdentifier)
+    UnresolvedAttribute.quotedString(exp.getUnresolvedAttribute.getUnparsedIdentifier)
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:400]: This is tricky... it's `CatalystSqlParser.parseMultipartIdentifier` v.s. `parseAttributeName` under the hood. Ideally `parseMultipartIdentifier` is more standard but following existing pyspark behavior is more important.

**Review by cloud-fan — APPROVED** [review]

**Review by amaliujia — APPROVED** [review]

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 6 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:400]: > This is tricky... it's `CatalystSqlParser.parseMultipartIdentifier` v.s. `parseAttributeName` under the hood. Ideally `parseMultipartIdentifier` is more standard but following existing pyspark behavior is more important.

This follows not just Pyspark but the Column API in Scala as well. There the string constructor uses exactly this method call.

File: python/pyspark/sql/tests/connect/test_connect_basic.py [hunk 7]
```diff
@@ -21,6 +21,7 @@
 
 import grpc  # type: ignore
 
+from pyspark.sql.connect.column import Column
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:24]: This line should be under `if have_pandas`. Here is a follow-up PR.
https://github.com/apache/spark/pull/38976

**Pull request closed without merging**